xdsl/traits.py

```diff
@@ -32,6 +33,23 @@ def verify(self, op: Operation) -> None:
 OpTraitInvT = TypeVar("OpTraitInvT", bound=OpTrait)
 
 
+class LazyTrait(OpTrait):
+    """A trait that postpone evaluation of other traits. Can be used to workaround circular dependency."""
+
+    parameters: tuple[Callable[[], OpTrait | tuple[OpTrait, ...]], ...]
+
+    def __init__(self, *parameters: Callable[[], OpTrait | tuple[OpTrait, ...]]):
+        super().__init__(parameters)
+
+    def verify(self, op: Operation) -> None:
+        for func in self.parameters:
+            traits = func()
```

Comment:
I think we might as well cache this to `self._traits`

Reply:
Cached in recent commit.